File: work/DSA/c++/recursion_2/binarySearch.cpp
#include <bits/stdc++.h>
using namespace std;

int helper(int input[], int end, int x, int start)
{
    if (start <= end)
    {
        int mid = (start + end) / 2;
        if (input[mid] == x)
        {
            return mid;
        }
        else if (input[mid] < x)
        {
            return helper(input, end, x, mid + 1);
        }
        else
        {
            return helper(input, mid - 1, x, start);
        }
    }
    else
    {
        return -1;
    }
}
int binarySearch(int input[], int length, int x)
{

    return helper(input, length - 1, x, 0);
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    int input[100000], length, element, ans;
    cin >> length;
    for (int i = 0; i < length; i++)
    {
        cin >> input[i];
        ;
    }

    cin >> element;
    ans = binarySearch(input, length, element);
    cout << ans << endl;
}

File: work/DSA/javascript/recursion_2/binarySearch.js
let input = [2, 3, 4, 5, 6, 8]
let x = 5

function helper(input, x, start, end) {
    if (start <= end) {
        let mid = Math.floor((start + end) / 2)
        if (input[mid] == x) {
            return mid
        }
        else if (input[mid] < x) {
            return helper(input, x, mid + 1, end)
        }
        else {
            return helper(input, x, start, mid - 1)
        }
    }
    return -1
}

function binarySearch(input, x, length) {
    return helper(input, x, 0, length - 1)
}

let answer = binarySearch(input, x, input.length)
console.log(answer)